File: Python/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num.replace(" ", "")
        print (self.card_num)
        pass
        

    def valid(self):
        validation = True
        card_number = list(self.card_num)

        if len(self.card_num) > 1 and self.card_num.isdigit():
            print(self.card_num.isdigit())
            i = 2
            sum = 0
            while i < len(card_number)+1:
                double = int(card_number[-i])*2
                print("The number is " + card_number[-i] + " and the double is " +  str(double))
                if double > 9:
                    double = double - 9
                    
                card_number[-i] = str(double)
                i += 2

            print (card_number)
            for item in card_number:
                sum = sum + int(item)
            print(sum)

            if not sum % 10 == 0:
                validation = False
            
        else:
            validation = False
            
        return validation

File: Ruby/luhn.rb
class Luhn
    def self.valid?(number)
      sum = 0
      number = number.gsub(' ','')
      
      if number.length < 2 or number.gsub(/[^0-9]/, '') != number then
        return false
      end
      # puts number.length
      for digit in (1..number.length) do
      #   puts number[digit-1]
        if digit%2 == 0  and number.length%2 != 0 then
          digit_result = 2 * number[digit-1].to_i
        elsif digit%2 != 0  and number.length%2 == 0 then
          digit_result = 2 * number[digit-1].to_i
        else
          digit_result = number[digit-1].to_i
        end
        if digit_result >= 10 then
          digit_result = digit_result - 9
        end
        sum += digit_result
        # puts sum 
      end
      
      return sum%10 == 0
    end
  end
    

puts Luhn.valid?("059")
puts Luhn.valid?("055 444 286")
puts Luhn.valid?("055 444 285")
puts Luhn.valid?("0")
puts Luhn.valid?("055-444-285")
puts Luhn.valid?("0000 0")